core/state-machine/src/trie_backend_essence.rs

```diff
@@ -141,6 +141,48 @@ impl<S: TrieBackendStorage<H>, H: Hasher> TrieBackendEssence<S, H> where H::Out:
 			debug!(target: "trie", "Error while iterating by prefix: {}", e);
 		}
 	}
+
+	/// Execute given closure for all keys starting with prefix. TODO fuse with for_keys_with_prefix
+	pub fn for_keys_with_child_prefix<F: FnMut(&[u8])>(&self, storage_key: &[u8], prefix: &[u8], mut f: F) {
+
+		let root: H::Out = match self.storage(storage_key) {
+			Ok(v) => v.map(|h|child_root_from_slice::<H>(&h[..])).unwrap_or(default_child_trie_root_hash::<H>(storage_key)),
+			Err(e) => {
+				debug!(target: "trie", "Error while iterating child storage: {}", e);
+				return;
+			}
+		};
+
+		let mut read_overlay = MemoryDB::default();
+		let eph = Ephemeral {
+			storage: &self.storage,
+			overlay: &mut read_overlay,
+		};
+
+		let mut iter = move || -> Result<(), Box<TrieError<H::Out>>> {
```

Comment:
I think this function has the same issue. We'd either:

* Remove support for variable length for multi-hash, that will make it so that we can use `H::Out` instead of `Vec<u8>` everywhere.
* Or take `root` as `Vec<u8>` here and move this iterator into `trie` struct.